src/gtc/passes/gtir_upcaster.py

```diff
@@ -43,6 +46,30 @@ def _update_node(node: Node, updated_children: Dict[str, TreeNode]) -> Expr:
         return node
 
 
+def _numpy_ufunc_upcasting_rule(*dtypes, ufunc: np.ufunc):
+    types = ufunc.types
+    for t in types:
+        inputs, output = t.split("->")
```

Comment:
I did not profile, but I see that it could be an issue. Especially repeatedly building the `table` in `op_to_ufunc`. I put some `@lru_cache` in places.